lwt/nbd_lwt_unix.ml

```diff
@@ -13,13 +13,76 @@
  *)
 open Nbd
 open Channel
-open Lwt
 
-let of_fd fd =
-  let read = Lwt_cstruct.(complete (read fd)) in
-  let write = Lwt_cstruct.(complete (write fd)) in
-  let close () = Lwt_unix.close fd in
-  { read; write; close }
+let return = Lwt.return
+let (>>=) = Lwt.(>>=)
+
+type tls_role =
+  | TlsClient of Ssl.context
+  | TlsServer of Ssl.context
+
+let debug_io = ref false
+
+(* TODO move this function into a library used by this and other executables *)
+(* Also the read/write/close functions that are inside tls_channel_of_fd *)
+let io_complete name offset op fd buffer =
+  if !debug_io
+  then Printf.fprintf stderr "%s offset=%s length=%d\n%!" name (match offset with Some x -> Int64.to_string x | None -> "None") (Cstruct.len buffer);
+  let ofs = buffer.Cstruct.off in
+  let len = buffer.Cstruct.len in
+  let buf = buffer.Cstruct.buffer in
+  let rec loop acc fd buf ofs len =
+    op fd buf ofs len >>= fun n ->
+    let len' = len - n in
+    let acc' = acc + n in
+    if len' = 0 || n = 0
+    then return acc'
+    else loop acc' fd buf (ofs + n) len' in
+  loop 0 fd buf ofs len >>= fun n ->
+  if n = 0 && len <> 0
+  then Lwt.fail End_of_file
+  else return ()
+
+let tls_channel_of_fd fd role () =
+  let ctx, ssl_start =
+    match role with
+      | TlsClient ctx -> ctx, Lwt_ssl.ssl_connect
+      | TlsServer ctx -> ctx, Lwt_ssl.ssl_accept
+  in
+  ssl_start fd ctx >>= fun sock ->
+  let offset = ref 0L in
+
+  let read_tls buf =
+    io_complete "read" (Some !offset) Lwt_ssl.read_bytes sock buf >>= fun () ->
+    offset := Int64.(add !offset (of_int (Cstruct.len buf)));
+    return () in
+
+  let write_tls buf =
+    io_complete "write" (Some !offset) Lwt_ssl.write_bytes sock buf >>= fun () ->
+    offset := Int64.(add !offset (of_int (Cstruct.len buf)));
+    return () in
+
+  let close_tls () =
+    ignore (Lwt_ssl.ssl_shutdown sock);
+    Lwt_ssl.close sock in
```

Comment:
We should test the NBD server with TLS connections as well to see if it leaks file descriptors.